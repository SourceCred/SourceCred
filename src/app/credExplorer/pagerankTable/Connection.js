// @flow

import React from "react";
import * as NullUtil from "../../../util/null";

import type {NodeAddressT} from "../../../core/graph";
import type {Connection} from "../../../core/attribution/graphToMarkovChain";
import type {ScoredConnection} from "../../../core/attribution/pagerankNodeDecomposition";
import type {DynamicPluginAdapter} from "../../pluginAdapter";

import {
  edgeVerb,
  nodeDescription,
  scoreDisplay,
  type SharedProps,
  type RowState,
} from "./shared";

type ConnectionRowListProps = {|
  +depth: number,
  +node: NodeAddressT,
  +sharedProps: SharedProps,
|};

export class ConnectionRowList extends React.PureComponent<
  ConnectionRowListProps
> {
  render() {
    const {depth, node, sharedProps} = this.props;
    const {pnd, maxEntriesPerList} = sharedProps;
    const {scoredConnections} = NullUtil.get(pnd.get(node));
    return (
      <React.Fragment>
        {scoredConnections
          .slice(0, maxEntriesPerList)
          .map((sc) => (
            <ConnectionRow
              key={JSON.stringify(sc.connection.adjacency)}
              depth={depth}
              target={node}
              scoredConnection={sc}
              sharedProps={sharedProps}
            />
          ))}
      </React.Fragment>
    );
  }
}

type ConnectionRowProps = {|
  +depth: number,
  +target: NodeAddressT,
  +scoredConnection: ScoredConnection,
  +sharedProps: SharedProps,
|};

export class ConnectionRow extends React.PureComponent<
  ConnectionRowProps,
  RowState
> {
  constructor() {
    super();
    this.state = {expanded: false};
  }
  render() {
    const {
      sharedProps,
      target,
      depth,
      scoredConnection: {connection, source, sourceScore, connectionScore},
    } = this.props;
    const {pnd, adapters} = sharedProps;
    const {expanded} = this.state;
    const {score: targetScore} = NullUtil.get(pnd.get(target));
    const connectionProportion = connectionScore / targetScore;
    const connectionPercent = (connectionProportion * 100).toFixed(2);

    return (
      <React.Fragment>
        <tr
          key="self"
          style={{backgroundColor: `rgba(0,143.4375,0,${1 - 0.9 ** depth})`}}
        >
          <td style={{display: "flex", alignItems: "flex-start"}}>
            <button
              style={{
                marginRight: 5,
                marginLeft: 15 * depth,
              }}
              onClick={() => {
                this.setState(({expanded}) => ({
                  expanded: !expanded,
                }));
              }}
            >
              {expanded ? "\u2212" : "+"}
            </button>
            <ConnectionView connection={connection} adapters={adapters} />
          </td>
          <td style={{textAlign: "right"}}>{connectionPercent}%</td>
          <td style={{textAlign: "right"}}>{scoreDisplay(sourceScore)}</td>
        </tr>
        {expanded && (
          <ConnectionRowList
            key="children"
            depth={depth + 1}
            node={source}
            sharedProps={sharedProps}
          />
        )}
      </React.Fragment>
    );
  }
}

export class ConnectionView extends React.PureComponent<{|
  +connection: Connection,
  +adapters: $ReadOnlyArray<DynamicPluginAdapter>,
|}> {
  render() {
    const {connection, adapters} = this.props;
    function Badge({children}) {
      return (
        // The outer <span> acts as a strut to ensure that the badge
        // takes up a full line height, even though its text is smaller.
        <span>
          <span
            style={{
              textTransform: "uppercase",
              fontWeight: 700,
              fontSize: "smaller",
            }}
          >
            {children}
          </span>
        </span>
      );
    }
    const {adjacency} = connection;
    switch (adjacency.type) {
      case "SYNTHETIC_LOOP":
        return <Badge>synthetic loop</Badge>;
      case "IN_EDGE":
        return (
          <span>
            <Badge>
              {edgeVerb(adjacency.edge.address, "BACKWARD", adapters)}
            </Badge>{" "}
            <span>{nodeDescription(adjacency.edge.src, adapters)}</span>
          </span>
        );
      case "OUT_EDGE":
        return (
          <span>
            <Badge>
              {edgeVerb(adjacency.edge.address, "FORWARD", adapters)}
            </Badge>{" "}
            <span>{nodeDescription(adjacency.edge.dst, adapters)}</span>
          </span>
        );
      default:
        throw new Error((adjacency.type: empty));
    }
  }
}
