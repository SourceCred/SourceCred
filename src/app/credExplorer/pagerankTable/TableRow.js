// @flow

import React, {type Node as ReactNode} from "react";

type TableRowProps = {|
  // How many layers of nested scope we're in (changes the color)
  +depth: number,
  // How many steps to indent the row (shifts button right)
  +indent: number,
  // The node that goes in the Description column
  +description: ReactNode,
  // What proportion should be formatted in the connection column
  +connectionProportion: ?number,
  // The score to format and display
  +score: number,
  // Children to show when the row is expanded
  +children: ReactNode,
  +showPadding: boolean,
|};
type TableRowState = {|
  expanded: boolean,
|};

export function scoreDisplay(score: number) {
  return score.toFixed(2);
}

export class TableRow extends React.PureComponent<
  TableRowProps,
  TableRowState
> {
  constructor() {
    super();
    this.state = {expanded: false};
  }
  render() {
    const {
      depth,
      indent,
      description,
      connectionProportion,
      score,
      children,
      showPadding,
    } = this.props;
    const {expanded} = this.state;
    const percent =
      connectionProportion == null
        ? ""
        : (connectionProportion * 100).toFixed(2) + "%";
    const backgroundColor = `hsla(150,100%,28%,${1 - 0.9 ** depth})`;
    return (
      <React.Fragment>
        {showPadding && <PaddingRow backgroundColor={backgroundColor} />}
        <tr key="self" style={{backgroundColor}}>
          <td style={{display: "flex", alignItems: "flex-start"}}>
            <button
              style={{
                marginRight: 5,
                marginLeft: 15 * indent + 5,
              }}
              onClick={() => {
                this.setState(({expanded}) => ({
                  expanded: !expanded,
                }));
              }}
            >
              {expanded ? "\u2212" : "+"}
            </button>
            {description}
          </td>
          <td style={{textAlign: "right"}}>{percent}</td>
          <td style={{textAlign: "right"}}>{scoreDisplay(score)}</td>
        </tr>
        {expanded && children}
        {showPadding && <PaddingRow backgroundColor={backgroundColor} />}
      </React.Fragment>
    );
  }
}

export function PaddingRow(props: {|+backgroundColor: string|}) {
  return (
    <tr
      style={{
        height: 12,
        backgroundColor: props.backgroundColor,
      }}
    >
      <td />
      <td />
      <td />
    </tr>
  );
}
