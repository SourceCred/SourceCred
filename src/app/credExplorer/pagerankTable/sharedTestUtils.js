// @flow

import {dynamicAdapterSet} from "../../adapters/demoAdapters";
import {pagerank} from "../../../core/attribution/pagerank";
import {defaultWeightsForAdapterSet} from "../weights/weights";

export const COLUMNS = () => ["Description", "", "Cred"];

export async function example() {
  const adapters = await dynamicAdapterSet();
  const weightedTypes = defaultWeightsForAdapterSet(adapters.static());
  const graph = adapters.graph();
  const pnd = await pagerank(graph, (_unused_Edge) => ({
    toWeight: 1,
    froWeight: 1,
  }));

  return {adapters, pnd, weightedTypes};
}
