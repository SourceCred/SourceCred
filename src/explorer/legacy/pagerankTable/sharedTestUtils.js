// @flow

import React from "react";

import {type NodeAddressT} from "../../../core/graph";
import {pagerank} from "../../../analysis/pagerank";
import {graph as demoGraph} from "../../../plugins/demo/graph";
import {declaration as demoDeclaration} from "../../../plugins/demo/declaration";
import type {SharedProps} from "./shared";

export const COLUMNS = () => ["Description", "", "Cred"];

export async function example() {
  const graph = demoGraph();
  const declarations = [demoDeclaration];
  const pnd = await pagerank(graph, (_unused_Edge) => ({
    forwards: 1,
    backwards: 1,
  }));
  const maxEntriesPerList = 123;
  const manualWeights: Map<NodeAddressT, number> = new Map();
  const onManualWeightsChange: (NodeAddressT, number) => void = jest.fn();
  const weightConfig: any = <div data-test-weight-config={true} />;
  const weightFileManager: any = <div data-test-weight-file-manager={true} />;

  const sharedProps: SharedProps = {
    graph,
    declarations,
    pnd,
    maxEntriesPerList,
    manualWeights,
    onManualWeightsChange,
  };

  return {
    pnd,
    maxEntriesPerList,
    sharedProps,
    manualWeights,
    onManualWeightsChange,
    weightConfig,
    weightFileManager,
  };
}
