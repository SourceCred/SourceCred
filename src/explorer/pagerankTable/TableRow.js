// @flow

import React, {type Node as ReactNode} from "react";
import {StyleSheet, css} from "aphrodite/no-important";

type TableRowProps = {|
  // How many layers of nested scope we're in (changes the color)
  +depth: number,
  // How many steps to indent the row (shifts button right)
  +indent: number,
  // The node that goes in the Description column
  +description: ReactNode,

  // The content for the "multiuse column"
  // Could be a weight slider or a cred proportion depending on context.
  +multiuseColumn: ReactNode,
  // The cred amount to format and display
  +cred: number,
  // Children to show when the row is expanded
  +children: ReactNode,
  +showPadding: boolean,
|};
type TableRowState = {|
  expanded: boolean,
|};

export function credDisplay(cred: number) {
  return cred.toFixed(2);
}

export class TableRow extends React.PureComponent<
  TableRowProps,
  TableRowState
> {
  constructor() {
    super();
    this.state = {expanded: false};
  }
  render() {
    const {
      depth,
      indent,
      description,
      cred,
      children,
      showPadding,
      multiuseColumn,
    } = this.props;
    const {expanded} = this.state;
    const backgroundColor = `hsla(150,100%,28%,${1 - 0.9 ** depth})`;
    const makeGradient = (color) =>
      `linear-gradient(to top, ${color}, ${color})`;
    const normalBackground = makeGradient(backgroundColor);
    const highlightBackground = makeGradient("#D8E1E8");
    const backgroundImage = `${normalBackground}, ${highlightBackground}`;

    return (
      <React.Fragment>
        {showPadding && <PaddingRow backgroundColor={backgroundColor} />}
        <tr
          key="self"
          style={{backgroundImage}}
          className={css(styles.hoverHighlight)}
        >
          <td style={{display: "flex", alignItems: "flex-start"}}>
            <button
              style={{
                marginRight: 5,
                marginLeft: 15 * indent + 5,
              }}
              onClick={() => {
                this.setState(({expanded}) => ({
                  expanded: !expanded,
                }));
              }}
            >
              {expanded ? "\u2212" : "+"}
            </button>
            {description}
          </td>
          <td style={{textAlign: "right"}}>{multiuseColumn}</td>
          <td style={{textAlign: "right"}}>
            <span style={{marginRight: 5}}>{credDisplay(cred)}</span>
          </td>
        </tr>
        {expanded && children}
        {showPadding && <PaddingRow backgroundColor={backgroundColor} />}
      </React.Fragment>
    );
  }
}

export function PaddingRow(props: {|+backgroundColor: string|}) {
  return (
    <tr
      style={{
        height: 12,
        backgroundColor: props.backgroundColor,
      }}
    >
      <td />
      <td />
      <td />
    </tr>
  );
}

const styles = StyleSheet.create({
  /* To apply 'hoverHighlight', provide a backgroundImage containing two <image>
 * data types (eg linear gradients). The first backgroundImage will be
 * the default background. The second backgroundImage will be applied on top
 * of the first background when the user hovers or tabs over the element.
 */

  hoverHighlight: {
    backgroundSize: "100% 100%, 0 0",
    ":hover": {
      backgroundSize: "100% 100%, 100% 100%",
    },
    ":focus-within": {
      backgroundSize: "100% 100%, 100% 100%",
    },
  },
});
