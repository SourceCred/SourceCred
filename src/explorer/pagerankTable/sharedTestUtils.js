// @flow

import React from "react";

import {type NodeAddressT} from "../../core/graph";
import {pagerank} from "../../analysis/pagerank";
import {dynamicExplorerAdapterSet} from "../../plugins/demo/explorerAdapter";
import type {SharedProps} from "./shared";

export const COLUMNS = () => ["Description", "", "Cred"];

export async function example() {
  const adapters = await dynamicExplorerAdapterSet();
  const graph = adapters.graph();
  const pnd = await pagerank(graph, (_unused_Edge) => ({
    toWeight: 1,
    froWeight: 1,
  }));
  const maxEntriesPerList = 123;
  const manualWeights: Map<NodeAddressT, number> = new Map();
  const onManualWeightsChange: (NodeAddressT, number) => void = jest.fn();
  const weightConfig: any = <div data-test-weight-config={true} />;

  const sharedProps: SharedProps = {
    adapters,
    pnd,
    maxEntriesPerList,
    manualWeights,
    onManualWeightsChange,
  };

  return {
    adapters,
    pnd,
    maxEntriesPerList,
    sharedProps,
    manualWeights,
    onManualWeightsChange,
    weightConfig,
  };
}
