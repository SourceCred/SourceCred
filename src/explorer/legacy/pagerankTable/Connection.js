// @flow

import React from "react";
import * as NullUtil from "../../../util/null";

import {type PluginDeclaration} from "../../../analysis/pluginDeclaration";
import {type NodeAddressT, Graph} from "../../../core/graph";
import type {Connection} from "../../../core/attribution/graphToMarkovChain";
import type {ScoredConnection} from "../../../analysis/pagerankNodeDecomposition";
import {TableRow} from "./TableRow";
import {NodeRow} from "./Node";

import {nodeDescription, edgeVerb, type SharedProps, Badge} from "./shared";

type ConnectionRowListProps = {|
  +depth: number,
  +node: NodeAddressT,
  +sharedProps: SharedProps,
  +connections: $ReadOnlyArray<ScoredConnection>,
|};

export class ConnectionRowList extends React.PureComponent<ConnectionRowListProps> {
  render() {
    const {depth, node, sharedProps, connections} = this.props;
    const {maxEntriesPerList} = sharedProps;
    return (
      <React.Fragment>
        {connections.slice(0, maxEntriesPerList).map((sc) => (
          <ConnectionRow
            key={JSON.stringify(sc.connection.adjacency)}
            depth={depth}
            target={node}
            scoredConnection={sc}
            sharedProps={sharedProps}
          />
        ))}
      </React.Fragment>
    );
  }
}

type ConnectionRowProps = {|
  +depth: number,
  +target: NodeAddressT,
  +scoredConnection: ScoredConnection,
  +sharedProps: SharedProps,
|};

export class ConnectionRow extends React.PureComponent<ConnectionRowProps> {
  render() {
    const {
      sharedProps,
      target,
      depth,
      scoredConnection: {connection, source, connectionScore},
    } = this.props;
    const {pnd, declarations, graph} = sharedProps;
    const {score: targetScore} = NullUtil.get(pnd.get(target));
    const connectionProportion = connectionScore / targetScore;
    const connectionPercent = (connectionProportion * 100).toFixed(2) + "%";

    const connectionView = (
      <ConnectionView
        connection={connection}
        declarations={declarations}
        graph={graph}
      />
    );
    return (
      <TableRow
        indent={2}
        depth={depth}
        description={connectionView}
        multiuseColumn={connectionPercent}
        showPadding={false}
        cred={connectionScore}
      >
        <NodeRow
          depth={depth + 1}
          showPadding={true}
          node={source}
          sharedProps={sharedProps}
        />
      </TableRow>
    );
  }
}

export class ConnectionView extends React.PureComponent<{|
  +connection: Connection,
  +declarations: $ReadOnlyArray<PluginDeclaration>,
  +graph: Graph,
|}> {
  render() {
    const {connection, declarations, graph} = this.props;
    const {adjacency} = connection;
    switch (adjacency.type) {
      case "SYNTHETIC_LOOP":
        return <Badge>synthetic loop</Badge>;
      case "IN_EDGE":
        return (
          <span>
            <Badge>
              {edgeVerb(adjacency.edge.address, "BACKWARD", declarations)}
            </Badge>{" "}
            <span>{nodeDescription(adjacency.edge.src, graph)}</span>
          </span>
        );
      case "OUT_EDGE":
        return (
          <span>
            <Badge>
              {edgeVerb(adjacency.edge.address, "FORWARD", declarations)}
            </Badge>{" "}
            <span>{nodeDescription(adjacency.edge.dst, graph)}</span>
          </span>
        );
      default:
        throw new Error((adjacency.type: empty));
    }
  }
}
