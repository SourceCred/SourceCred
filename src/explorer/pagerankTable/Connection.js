// @flow

import React from "react";
import * as NullUtil from "../../util/null";

import type {NodeAddressT} from "../../core/graph";
import type {Connection} from "../../core/attribution/graphToMarkovChain";
import type {ScoredConnection} from "../../analysis/pagerankNodeDecomposition";
import {DynamicExplorerAdapterSet} from "../adapters/explorerAdapterSet";
import {TableRow} from "./TableRow";
import {NodeRow} from "./Node";

import {edgeVerb, nodeDescription, type SharedProps, Badge} from "./shared";

type ConnectionRowListProps = {|
  +depth: number,
  +node: NodeAddressT,
  +sharedProps: SharedProps,
  +connections: $ReadOnlyArray<ScoredConnection>,
|};

export class ConnectionRowList extends React.PureComponent<ConnectionRowListProps> {
  render() {
    const {depth, node, sharedProps, connections} = this.props;
    const {maxEntriesPerList} = sharedProps;
    return (
      <React.Fragment>
        {connections.slice(0, maxEntriesPerList).map((sc) => (
          <ConnectionRow
            key={JSON.stringify(sc.connection.adjacency)}
            depth={depth}
            target={node}
            scoredConnection={sc}
            sharedProps={sharedProps}
          />
        ))}
      </React.Fragment>
    );
  }
}

type ConnectionRowProps = {|
  +depth: number,
  +target: NodeAddressT,
  +scoredConnection: ScoredConnection,
  +sharedProps: SharedProps,
|};

export class ConnectionRow extends React.PureComponent<ConnectionRowProps> {
  render() {
    const {
      sharedProps,
      target,
      depth,
      scoredConnection: {connection, source, connectionScore},
    } = this.props;
    const {pnd, adapters} = sharedProps;
    const {score: targetScore} = NullUtil.get(pnd.get(target));
    const connectionProportion = connectionScore / targetScore;
    const connectionPercent = (connectionProportion * 100).toFixed(2) + "%";

    const connectionView = (
      <ConnectionView connection={connection} adapters={adapters} />
    );
    return (
      <TableRow
        indent={2}
        depth={depth}
        description={connectionView}
        multiuseColumn={connectionPercent}
        showPadding={false}
        cred={connectionScore}
      >
        <NodeRow
          depth={depth + 1}
          showPadding={true}
          node={source}
          sharedProps={sharedProps}
        />
      </TableRow>
    );
  }
}

export class ConnectionView extends React.PureComponent<{|
  +connection: Connection,
  +adapters: DynamicExplorerAdapterSet,
|}> {
  render() {
    const {connection, adapters} = this.props;
    const {adjacency} = connection;
    switch (adjacency.type) {
      case "SYNTHETIC_LOOP":
        return <Badge>synthetic loop</Badge>;
      case "IN_EDGE":
        return (
          <span>
            <Badge>
              {edgeVerb(adjacency.edge.address, "BACKWARD", adapters)}
            </Badge>{" "}
            <span>{nodeDescription(adjacency.edge.src, adapters)}</span>
          </span>
        );
      case "OUT_EDGE":
        return (
          <span>
            <Badge>
              {edgeVerb(adjacency.edge.address, "FORWARD", adapters)}
            </Badge>{" "}
            <span>{nodeDescription(adjacency.edge.dst, adapters)}</span>
          </span>
        );
      default:
        throw new Error((adjacency.type: empty));
    }
  }
}
