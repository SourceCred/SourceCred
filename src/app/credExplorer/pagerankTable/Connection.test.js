// @flow

import React from "react";
import {shallow} from "enzyme";
import * as NullUtil from "../../../util/null";

import type {Connection} from "../../../core/attribution/graphToMarkovChain";
import {ConnectionRowList, ConnectionRow, ConnectionView} from "./Connection";
import {example} from "./sharedTestUtils";
import {TableRow} from "./TableRow";

require("../../testUtil").configureEnzyme();

describe("app/credExplorer/pagerankTable/Connection", () => {
  beforeEach(() => {
    // $ExpectFlowError
    console.error = jest.fn();
    // $ExpectFlowError
    console.warn = jest.fn();
  });
  afterEach(() => {
    expect(console.warn).not.toHaveBeenCalled();
    expect(console.error).not.toHaveBeenCalled();
  });

  describe("ConnectionRowList", () => {
    async function setup(maxEntriesPerList: number = 100000) {
      const {adapters, pnd, nodes} = await example();
      const depth = 2;
      const node = nodes.bar1;
      const sharedProps = {adapters, pnd, maxEntriesPerList};
      const component = (
        <ConnectionRowList
          depth={depth}
          node={node}
          sharedProps={sharedProps}
        />
      );
      const element = shallow(component);
      return {element, depth, node, sharedProps};
    }
    it("creates `ConnectionRow`s with the right props", async () => {
      const {element, depth, node, sharedProps} = await setup();
      const connections = NullUtil.get(sharedProps.pnd.get(node))
        .scoredConnections;
      const rows = element.find("ConnectionRow");
      expect(rows).toHaveLength(connections.length);
      const rowPropses = rows.map((row) => row.props());
      // Order should be the same as the order in the decomposition.
      expect(rowPropses).toEqual(
        connections.map((sc) => ({
          depth,
          sharedProps,
          target: node,
          scoredConnection: sc,
        }))
      );
    });
    it("limits the number of rows by `maxEntriesPerList`", async () => {
      const maxEntriesPerList = 1;
      const {element, node, sharedProps} = await setup(maxEntriesPerList);
      const connections = NullUtil.get(sharedProps.pnd.get(node))
        .scoredConnections;
      expect(connections.length).toBeGreaterThan(maxEntriesPerList);
      const rows = element.find("ConnectionRow");
      expect(rows).toHaveLength(maxEntriesPerList);
      const rowConnections = rows.map((row) => row.prop("scoredConnection"));
      // Should have selected the right nodes.
      expect(rowConnections).toEqual(connections.slice(0, maxEntriesPerList));
    });
  });

  describe("ConnectionRow", () => {
    async function setup() {
      const {pnd, adapters, nodes} = await example();
      const sharedProps = {adapters, pnd, maxEntriesPerList: 123};
      const target = nodes.bar1;
      const {scoredConnections} = NullUtil.get(pnd.get(target));
      const alphaConnections = scoredConnections.filter(
        (sc) => sc.source === nodes.fooAlpha
      );
      expect(alphaConnections).toHaveLength(1);
      const scoredConnection = alphaConnections[0];
      const depth = 2;
      const component = (
        <ConnectionRow
          depth={depth}
          target={target}
          scoredConnection={scoredConnection}
          sharedProps={sharedProps}
        />
      );
      const row = shallow(component).find(TableRow);
      return {row, depth, target, scoredConnection, sharedProps};
    }
    describe("instantiates a TableRow", () => {
      it("with the correct depth", async () => {
        const {row, depth} = await setup();
        expect(row.props().depth).toBe(depth);
      });
      it("with the sourceScore", async () => {
        const {row, scoredConnection} = await setup();
        expect(row.props().score).toBe(scoredConnection.sourceScore);
      });
      it("with the connectionProportion", async () => {
        const {row, target, scoredConnection, sharedProps} = await setup();
        const targetScore = NullUtil.get(sharedProps.pnd.get(target)).score;
        expect(row.props().connectionProportion).toBe(
          scoredConnection.connectionScore / targetScore
        );
      });
      it("with a ConnectionView as description", async () => {
        const {row, sharedProps, scoredConnection} = await setup();
        const {adapters} = sharedProps;
        const description = row.props().description;
        const cv = shallow(description).instance();
        expect(cv).toBeInstanceOf(ConnectionView);
        expect(cv.props.connection).toEqual(scoredConnection.connection);
        expect(cv.props.adapters).toEqual(adapters);
      });
      describe("with a ConnectionRowList as children", () => {
        function getChildren(row) {
          const children = row.props().children;
          return shallow(children).instance();
        }
        it("which is a ConnectionRowList", async () => {
          const {row} = await setup();
          expect(getChildren(row)).toBeInstanceOf(ConnectionRowList);
        });
        it("which has incremented depth", async () => {
          const {row, depth} = await setup();
          expect(getChildren(row).props.depth).toBe(depth + 1);
        });
        it("which has the connection source as its node target", async () => {
          const {row, scoredConnection} = await setup();
          expect(getChildren(row).props.node).toBe(scoredConnection.source);
        });
        it("which has the right sharedProps", async () => {
          const {row, sharedProps} = await setup();
          expect(getChildren(row).props.sharedProps).toBe(sharedProps);
        });
      });
    });
  });
  describe("ConnectionView", () => {
    async function setup() {
      const {pnd, adapters, nodes} = await example();
      const {scoredConnections} = NullUtil.get(pnd.get(nodes.bar1));
      const connections = scoredConnections.map((sc) => sc.connection);
      function connectionByType(t) {
        return NullUtil.get(
          connections.filter((c) => c.adjacency.type === t)[0],
          `Couldn't find connection for type ${t}`
        );
      }
      const inConnection = connectionByType("IN_EDGE");
      const outConnection = connectionByType("OUT_EDGE");
      const syntheticConnection = connectionByType("SYNTHETIC_LOOP");
      function cvForConnection(connection: Connection) {
        return shallow(
          <ConnectionView adapters={adapters} connection={connection} />
        );
      }
      return {
        adapters,
        connections,
        pnd,
        cvForConnection,
        inConnection,
        outConnection,
        syntheticConnection,
      };
    }
    it("always renders exactly one `Badge`", async () => {
      const {
        cvForConnection,
        inConnection,
        outConnection,
        syntheticConnection,
      } = await setup();
      for (const connection of [
        syntheticConnection,
        inConnection,
        outConnection,
      ]) {
        expect(cvForConnection(connection).find("Badge")).toHaveLength(1);
      }
    });
    it("for inward connections, renders a `Badge` and description", async () => {
      const {cvForConnection, inConnection} = await setup();
      const view = cvForConnection(inConnection);
      const outerSpan = view.find("span").first();
      const badge = outerSpan.find("Badge");
      const description = outerSpan.children().find("span");
      expect(badge.children().text()).toEqual("is barred by");
      expect(description.text()).toEqual('bar: NodeAddress["bar","a","1"]');
    });
    it("for outward connections, renders a `Badge` and description", async () => {
      const {cvForConnection, outConnection} = await setup();
      const view = cvForConnection(outConnection);
      const outerSpan = view.find("span").first();
      const badge = outerSpan.find("Badge");
      const description = outerSpan.children().find("span");
      expect(badge.children().text()).toEqual("bars");
      expect(description.text()).toEqual("xox node!");
    });
    it("for synthetic connections, renders only a `Badge`", async () => {
      const {cvForConnection, syntheticConnection} = await setup();
      const view = cvForConnection(syntheticConnection);
      expect(view.find("span")).toHaveLength(0);
      expect(
        view
          .find("Badge")
          .children()
          .text()
      ).toEqual("synthetic loop");
    });
  });
});
