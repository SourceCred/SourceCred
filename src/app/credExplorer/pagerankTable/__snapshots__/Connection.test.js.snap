// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/credExplorer/pagerankTable/Connection ConnectionRow has proper depth-based styling 1`] = `
Object {
  "buttonStyle": Object {
    "marginLeft": 0,
    "marginRight": 5,
  },
  "colorDepth": 0,
  "indentDepth": 0,
  "trStyle": Object {
    "backgroundColor": "rgba(0,143.4375,0,0)",
  },
}
`;

exports[`app/credExplorer/pagerankTable/Connection ConnectionRow has proper depth-based styling 2`] = `
Object {
  "buttonStyle": Object {
    "marginLeft": 15,
    "marginRight": 5,
  },
  "colorDepth": 0,
  "indentDepth": 1,
  "trStyle": Object {
    "backgroundColor": "rgba(0,143.4375,0,0)",
  },
}
`;

exports[`app/credExplorer/pagerankTable/Connection ConnectionRow has proper depth-based styling 3`] = `
Object {
  "buttonStyle": Object {
    "marginLeft": 0,
    "marginRight": 5,
  },
  "colorDepth": 1,
  "indentDepth": 0,
  "trStyle": Object {
    "backgroundColor": "rgba(0,143.4375,0,0.09999999999999998)",
  },
}
`;

exports[`app/credExplorer/pagerankTable/Connection ConnectionRow has proper depth-based styling 4`] = `
Object {
  "buttonStyle": Object {
    "marginLeft": 15,
    "marginRight": 5,
  },
  "colorDepth": 1,
  "indentDepth": 1,
  "trStyle": Object {
    "backgroundColor": "rgba(0,143.4375,0,0.09999999999999998)",
  },
}
`;
